### PR TITLE
[Extend Governor APIs] Extensions and ERD schemas

### PR DESCRIPTION
**Pull request by bailinhe**
Currently Governor API supports addons by allowing standalone applications to listen to various governor events and perform additional tasks accordingly. This approach, while being simple yet powerful and flexible, limits the capabilities of the addons. Namely, addons that provide configurability for the end users, require communications between the users and the addons through the Governor API.

This **[Extend Governor APIs]** series PRs aim to provide an alternative way, in addition to the existing addons, to improve the Governor API's extendibility, by implementing a solution that allows the API itself to be extended, in a design that is heavily influenced by Kubernetes' CRD and operator patten.

Data schema created:

- [x] Extensions
- [x] Extension Resource Definitions

```mermaid
erDiagram
  extension {
    uuid id
    string url
    string name
    string slug
    string description
    bool enabled
    enum status "offline, online"
  }

  extension_resource_definition {
    uuid id
    string name
    string slug_singular
    string slug_plural
    enum scope "system or user"

    bool enabled
    string version
    jsonb schema "JSON schema"
  }

  user_extension_resource {
    uuid id
    jsonb resource
  }

  system_extension_resource {
    uuid id
    jsonb resource
  }

  extension ||--o{ extension_resource_definition: has

  user ||--o{ user_extension_resource: has
  extension_resource_definition ||--o{ user_extension_resource: validates

  extension_resource_definition ||--o{ system_extension_resource: validates
```

### CODE REVIEW

File: db/migrations/00034_api_extensions.sql [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+CREATE TYPE extension_status AS ENUM('online', 'offline');
+CREATE TYPE extension_resource_scope AS ENUM('user', 'system');
+
+CREATE TABLE extensions (
+  id UUID PRIMARY KEY NOT NULL DEFAULT gen_random_uuid(),
+  name STRING NOT NULL,
+  description STRING NOT NULL,
+  enabled BOOL NOT NULL,
+  slug STRING NOT NULL,
+	status extension_status NOT NULL DEFAULT 'offline',
```
Comment (fishnix) [inline db/migrations/00034_api_extensions.sql:12]: whitespace?

File: db/migrations/00034_api_extensions.sql [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+);
+
+CREATE TABLE extension_resource_definitions (
+  id UUID PRIMARY KEY NOT NULL DEFAULT gen_random_uuid(),
+  name STRING NOT NULL,
+  description STRING NOT NULL,
+  enabled BOOL NOT NULL DEFAULT true,
+
+  slug_singular STRING NOT NULL,
+  slug_plural STRING NOT NULL,
```
Comment (fishnix) [inline db/migrations/00034_api_extensions.sql:30]: do we need to store two slugs?

File: db/migrations/00034_api_extensions.sql [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  version STRING NOT NULL,
+	scope extension_resource_scope NOT NULL,
+	schema JSONB NOT NULL,
```
Comment (fishnix) [inline db/migrations/00034_api_extensions.sql:33]: 👀

Comment (fishnix) [reply db/migrations/00034_api_extensions.sql:33]: and below

File: db/migrations/00034_api_extensions.sql [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  enabled BOOL NOT NULL,
+  slug STRING NOT NULL,
+	status extension_status NOT NULL DEFAULT 'offline',
+
+  created_at TIMESTAMPTZ NOT NULL DEFAULT now(),
+  updated_at TIMESTAMPTZ NOT NULL DEFAULT now(),
+  deleted_at TIMESTAMPTZ NULL,
+
+  CONSTRAINT extension_slug_key UNIQUE (slug) WHERE deleted_at IS NULL,
+  INDEX (slug, deleted_at) STORING (name, description, enabled, status, created_at, updated_at),
```
Comment (bailinhe) [inline db/migrations/00034_api_extensions.sql:19]: Indices are suggested by cockroachDB when `explain` against the most common queries

**Review by fishnix — APPROVED** [review]
couple minor comments 👍

File: db/migrations/00034_api_extensions.sql (hunk 2 above)
Comment (bailinhe) [reply db/migrations/00034_api_extensions.sql:30]: we will use the singular to identify the ERD itself, and use the plural for the URIs of the extension resources

Comment (fishnix) [reply db/migrations/00034_api_extensions.sql:30]: but cant we just generate one from the other? or is it hard bc its been slug-ified already?

**Review by fishnix — APPROVED** [review]

File: db/migrations/00034_api_extensions.sql (hunk 2 above)
Comment (bailinhe) [reply db/migrations/00034_api_extensions.sql:30]: yeah but we also have `goose` vs `geese` and `moose` vs `moose`, I'd say the extension developer should take care of this

Comment (fishnix) [reply db/migrations/00034_api_extensions.sql:30]: thats fine with me - there are actually libraries to do that i think?

**Pull request merged**